api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,93 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
+// This structure allows for traversal to happen as data is becoming available, allowing for optimized matching.
+//
+// We talk about an "associated operation" throughout this docs: this is the context dependent operation that is configured with this match tree. For example, this could be a HTTP/TCP filter.
+message MatchTree {
+  // The action that should be taken upon matching.
+  message MatchAction {
+    oneof action {
+      option (validate.required) = true;
+
+      // The associated operation should be skipped.
+      bool skip = 1 [(validate.rules).bool = {const: true}];
+
+      // The specified callback value should be passed to the associated operation.
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    message LeafMatcher {
+      oneof matcher {
+        option (validate.required) = true;
+
+        MatchPredicate predicate = 1;
+
+        // [#not-implemented-hide:]
+        core.v3.TypedExtensionConfig typed_config = 2;
+      }
+
+      // The action to perform if the match evaluates to true.
+      MatchAction action = 3 [(validate.rules).message = {required: true}];
+    }
+
+    // A linear set of match conditions that will be evaluated in order.
+    // Each matcher will be attempted in order until one succeeds.
+    // If none of the matchers match, the no_match_action will be performed if specified.
+    repeated LeafMatcher matchers = 1 [(validate.rules).repeated = {min_items: 1}];
+
+    // Action to perform if none of the above matchers match.
+    MatchAction no_match_action = 3;
+  }
+
+  message SublinearMatcher {
+    message MultimapMatcher {
+      // The key to apply this matcher for.
+      string key = 1 [(validate.rules).string = {min_len: 1}];
+
+      // The namespace of this key. This is used to distinguish multiple maps
+      // that could be consulted during matching.
+      string key_namespace = 2 [(validate.rules).string = {min_len: 1}];
+
+      // Mapping from exact value matches to the matching subtree to use when
+      // the match applies.
+      // If a header value matches both a prefix and an exact match, the exact
+      // match path will be taken.
+      map<string, MatchTree> exact_matches = 3;
+
+      // Mapping from prefix matches to the matching subtree to use when the match applies.
+      // If a header value matches both a prefix and an exact match, the exact match path will be taken.
+      // If multiple prefixes match, the longest prefix path will be taken.
+      map<string, MatchTree> prefix_matches = 4;
+    }
+
+    oneof matcher {
+      option (validate.required) = true;
+
+      MultimapMatcher multimap_matcher = 1;
+
+      // [#not-implemented-hide:]
+      core.v3.TypedExtensionConfig typed_config = 2;
+    }
+
+    // If none of the above matchers match, attempt to match against this subtree.
+    MatchTree no_match_tree = 3;
```

Comment:
Open question: should this be a `MatchAction`? Does making this recursive make things too flexible?

Reply:
My vote is to just make this `MatchAction` unless there is a good reason to do so.

Reply:
How do we build logic like "use exact_match for majority of routes, then do the remaining via expensive regular expression"?
Arguably for HTTP scenario specifically it can be solved by "use exact match for majority of routes, then do the rest of selection in "/" prefix match's tree" (so that `no_match` is never used). It's semantically the same but via some http-specific workarounds.

What issues do you have in mind by making `no_match` special case (using action vs tree)? And which of this issues are effectively invalidated by using `"/" prefix` workaround?

Reply:
If there is a good reason to keep it as a tree (per @euroelessar) no objection from me.